internal/component/prometheus/remote/queue/component.go

```diff
@@ -0,0 +1,160 @@
+	"context"
+	"path/filepath"
+	"reflect"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/alloy/internal/component"
+	"github.com/grafana/alloy/internal/component/prometheus/remote/queue/filequeue"
+	"github.com/grafana/alloy/internal/component/prometheus/remote/queue/network"
+	"github.com/grafana/alloy/internal/component/prometheus/remote/queue/serialization"
+	"github.com/grafana/alloy/internal/component/prometheus/remote/queue/types"
+	"github.com/grafana/alloy/internal/featuregate"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/storage"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:      "prometheus.remote.queue",
+		Args:      Arguments{},
+		Exports:   Exports{},
+		Stability: featuregate.StabilityExperimental,
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return NewComponent(opts, args.(Arguments))
+		},
+	})
+}
+
+func NewComponent(opts component.Options, args Arguments) (*Queue, error) {
+	s := &Queue{
+		opts:      opts,
+		args:      args,
+		log:       opts.Logger,
+		endpoints: map[string]*endpoint{},
+	}
+	s.opts.OnStateChange(Exports{Receiver: s})
+	err := s.createEndpoints()
+	for _, ep := range s.endpoints {
+		ep.Start()
+	}
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+// Queue is a queue based WAL used to send data to a remote_write endpoint. Queue supports replaying
+// and TTLs.
+type Queue struct {
+	mut       sync.RWMutex
+	args      Arguments
+	opts      component.Options
+	log       log.Logger
+	endpoints map[string]*endpoint
+}
+
+// Run starts the component, blocking until ctx is canceled or the component
+// suffers a fatal error. Run is guaranteed to be called exactly once per
+// Component.
+func (s *Queue) Run(ctx context.Context) error {
+	defer func() {
+		s.mut.Lock()
+		defer s.mut.Unlock()
+
+		for _, ep := range s.endpoints {
+			ep.Stop()
+		}
+	}()
+
+	<-ctx.Done()
+	return nil
+}
+
+// Update provides a new Config to the component. The type of newConfig will
+// always match the struct type which the component registers.
+//
+// Update will be called concurrently with Run. The component must be able to
+// gracefully handle updating its config while still running.
+//
+// An error may be returned if the provided config is invalid.
+func (s *Queue) Update(args component.Arguments) error {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+
+	newArgs := args.(Arguments)
+	sync.OnceFunc(func() {
+		s.opts.OnStateChange(Exports{Receiver: s})
+	})
+	// If they are the same do nothing.
+	if reflect.DeepEqual(newArgs, s.args) {
+		return nil
+	}
+	s.args = newArgs
+	// TODO @mattdurham need to cycle through the endpoints figuring out what changed instead of this global stop and start.
+	// TODO @mattdurham is there an issue/race condition with stopping these while the appender is still going on.
+	if len(s.endpoints) > 0 {
+		for _, ep := range s.endpoints {
+			ep.Stop()
```

Comment:
No, it will in general call the cancel on the context, which will cascade down. You will in general lose any data in the queues. Will add documentation.

Reply:
Any plans on adding a drain behaviour? Could perhaps include that in the comment.  I anticipate someone opening an issue about metric gaps one day.

Reply:
One of the original versions I did had it but it was a bit fiddly. Its something I def want to do but removed it from scope for the first iteration; like specific TLS settings, auto scaling the number of parallel queues  and a few other things.